admin/failures/overview/endpoint.js

```diff
@@ -28,9 +37,10 @@ module.exports = (purchases) => (req, res) => {
             })
             .map((o) => {
                 const latestError = o.data.errors && o.data.errors.length && o.data.errors[o.data.errors.length - 1];
+                const latestErrorAt = latestError ? prettifyDate(latestError.at) : prettifyDate(o.data.viewModel.date);

```

Comment:
Good case for an extraction here, too. You can even do a nice trick by doing a general function:

```js
getFirstPrettyDate(latestError?.at, o.data.viewModel.date);

// where...

function getFirstPrettyDate(...dateStrings) {
    for(const dateString of dateStrings) {
        if(dateString) {
            return prettyDate(dateString);
        }
    }
}
```

Notice the rest parameters syntax (`...dateStrings`): https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Functions/rest_parameters

Reply:
In general, a function should do one thing on one level of abstraction.

In this case we have two levels of abstraction mixed:

- Deciding where to get a date from, based on which data exists (relatively high-level business logic decision)
- Transforming a date into the correct display format

When it is in a simple teritary we mix the two decisions. My suggestion is that we create a general function that returns a date from several prioritized possibilities (*without* knowing the shape or purpose of our data!) and a higher level of abstraction making the decision on data and simply passing it to a function.

This splits the two levels of abstraction and makes it easier to reason about.